### PR TITLE
adding new overloads to IOUtils with Path for some file only methods

### LINKED ISSUE #1294: Add Path versions of IOUtil openGzipFileForWriting
IOUtil doesn't have Path equivalents of openXForWriting.  It should.

### PR DESCRIPTION
**Pull request by lbergelson**
### Description
Add Path overloads of some functions in IOUtil that were missing them.
Fixes #1294
### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/samtools/util/IOUtil.java [hunk 1]
```diff
@@ -672,30 +668,46 @@ public static InputStream openGzipFileForReading(final Path path) {
...
+     */
+    public static OutputStream openFileForWriting(final Path path, OpenOption... openOptions) {
         try {
-            if (file.getName().endsWith(".gz") ||
-                file.getName().endsWith(".bfq")) {
-                return openGzipFileForWriting(file, append);
-            }
-            else {
-                return new FileOutputStream(file, append);
+            if(hasGzipFileExtension(path)) {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/IOUtil.java]:
```suggestion
            if (hasGzipFileExtension(path)) {
```

File: src/test/java/htsjdk/samtools/util/IOUtilTest.java [hunk 2]
```diff
@@ -582,6 +582,25 @@ public void testCompressionLevel(final Path file, final String extension, final
...
+
+    @Test(dataProvider = "getExtensions")
+    public void testReadWriteJimfs(String extension) throws IOException {
+        final Path jmfsRoot = inMemoryFileSystem.getRootDirectories().iterator().next();
+        final Path tmp = Files.createTempFile(jmfsRoot, "test", extension);
+        final String expected = "lorem ipswitch, nantucket, bucket";
+        try( final Writer out = IOUtil.openFileForBufferedWriting(tmp)){
+            out.write(expected);
+        }
+        try( final BufferedReader in = IOUtil.openFileForBufferedReading(tmp)){
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/IOUtilTest.java]: can you somehow check that the zipped files are **really** zipped and that the plaintext ones are really plaintext?

Comment (lbergelson) [reply src/test/java/htsjdk/samtools/util/IOUtilTest.java]: done

**Comment by lbergelson** [comment]
@yfarjoun I responded to your comments, good to go?

File: src/test/java/htsjdk/samtools/util/IOUtilTest.java [hunk 3]
```diff
@@ -582,6 +582,33 @@ public void testCompressionLevel(final Path file, final String extension, final
...
+                {".bfq", true},
+                {".txt", false}};
+    }
+
+    @Test(dataProvider = "getExtensions")
+    public void testReadWriteJimfs(String extension, boolean gzipped) throws IOException {
+        final Path jmfsRoot = inMemoryFileSystem.getRootDirectories().iterator().next();
+        final Path tmp = Files.createTempFile(jmfsRoot, "test", extension);
+        final String expected = "lorem ipswitch, nantucket, bucket";
+        try( final Writer out = IOUtil.openFileForBufferedWriting(tmp)){
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/IOUtilTest.java]:
```suggestion
        try (Writer out = IOUtil.openFileForBufferedWriting(tmp)){
```

File: src/test/java/htsjdk/samtools/util/IOUtilTest.java [hunk 4]
```diff
@@ -582,6 +582,33 @@ public void testCompressionLevel(final Path file, final String extension, final
...
+            out.write(expected);
+        }
+
+        try( final InputStream in = new BufferedInputStream(Files.newInputStream(tmp))){
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/IOUtilTest.java]:
```suggestion
        try (InputStream in = new BufferedInputStream(Files.newInputStream(tmp))){
```

File: src/test/java/htsjdk/samtools/util/IOUtilTest.java [hunk 5]
```diff
@@ -582,6 +582,33 @@ public void testCompressionLevel(final Path file, final String extension, final
...
+               Assert.assertEquals(IOUtil.isGZIPInputStream(in), gzipped);
+        }
+
+        try( final BufferedReader in = IOUtil.openFileForBufferedReading(tmp)){
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/IOUtilTest.java]:
```suggestion
        try (BufferedReader in = IOUtil.openFileForBufferedReading(tmp)){
```

**Pull request merged**